IntelBluetoothFirmware/IntelBluetoothFirmware.cpp

```diff
@@ -1169,14 +1189,41 @@ void IntelBluetoothFirmware::onLoadFW(OSKextRequestTag requestTag, OSReturn resu
 
 IOReturn IntelBluetoothFirmware::setPowerState(unsigned long powerStateOrdinal, IOService *whatDevice)
 {
-//    XYLog("setPowerState powerStateOrdinal=%lu\n", powerStateOrdinal);
+    XYLog("setPowerState(0x%lx)\n", powerStateOrdinal);
+
+    if (powerStateOrdinal < mDevicePowerState) {
+        mDevicePowerState = powerStateOrdinal;
+    }
+
+    switch (powerStateOrdinal) {
+        case kMyOffPowerState:
+            // Wait for outstanding IO to complete before putting device into a
+            // lowe power state
+            IOLockLock(mLock);
+            while (mOutstandIO != 0) {
+                IOLockSleep(mLock, &mOutstandIO, THREAD_UNINT);
+            }
+            IOLockUnlock(mLock);
+            break;
+    }
+
+    if (powerStateOrdinal > mDevicePowerState) {
+        mDevicePowerState = powerStateOrdinal;
+    }
     return IOPMAckImplied;
 }
 
+void IntelBluetoothFirmware::powerChangeDone(unsigned long stateNumber)
+{
+    // Wake up threads waiting on power state change
+    XYLog("powerChangeDone(0x%lx)\n", stateNumber);
+    IOLockWakeup(mLock, &mDevicePowerState, false);
+}
+
 void IntelBluetoothFirmware::stop(IOService *provider)
```

Comment:
what are the meaning of `setPowerState` and `powerChangeDone` you are doing here, why do you confirm it can fix the wakeup problem?

Reply:
I suspect that the Bluetooth device becomes unavailable is due to It doesn't wait for the device power state is being changed done.
So I force it to wait until the power state is being changed done.
It seems to work fine on my LG-Gram but as you can see, there is someone feedback it doesn't work.

Reply:
I think the wakeup issus is not come from here... the firmware successfully uploaded but didn't work, so I guess the kext may work well. I will take sometime to try another way tonight.